### DISCUSSION TITLE
"Transitive" language interoperability

### DISCUSSION THREAD

**Discussion by mallman** [carbon-language/carbon-lang#1742] [Language design]
The Carbon project includes bidirectional compatibility with C++ as a goal. I have a related question.

Hypothetically, if programming language X has C++ interoperability, will X—by design and implementation of Carbon—automatically have Carbon interoperability? This would be a nice property for any "C++ successor".

Non-hypothetically, Swift has a project/goal to enable, at least unidirectional if not bidirectional, C++ interoperability. I don't have details of that project—I'm not part of the effort. I'm just a Swift programmer who wants to use C++ libraries without hassle.

**Comment by jonmeow** [comment]
> Hypothetically, if programming language X has C++ interoperability, will X—by design and implementation of Carbon—automatically have Carbon interoperability? This would be a nice property for any "C++ successor".

This will depend on how the language implements its interoperability. It's something I'd assume is likely to work but is impossible to broadly promise. e.g., suppose Carbon supports generating C++ headers for some endpoints (which personally I think is likely for cross C++  compiler interop): even if language X supports interop, it could still be really unhappy with Carbon's generated headers.
